File: tif_customization/tif_customization/page/accounts_status/accounts_status.js
const TIF_ACCT_PALETTE = {
	lines2: ["#2563eb", "#dc2626"],
	donut: ["#4f46e5", "#ea580c", "#059669", "#e11d48", "#7c3aed", "#0891b2"],
};

frappe.pages["accounts-status"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: "Accounts Status",
		single_column: true,
	});

	class TIFAccountsStatusDashboard {
		constructor(page) {
			this.page = page;
			this.apex_loaded = false;
			this.charts = {};
			this.filters = {
				company: frappe.defaults.get_user_default("Company") || "",
				from_date: null,
				to_date: frappe.datetime.get_today(),
			};
		}

		async make() {
			this.render();
			this.inject_styles();
			this.make_filters();
			await this.load_apexcharts();
			await this.refresh();
		}

		render() {
			this.page.main.html(`
				<div class="tif-acctdash">
					<p class="text-muted" style="font-size:12px;margin:0 0 10px">
						<strong>Trustee monitoring</strong> — receivables, payables, draft vouchers, income/expense and bank/cash balances.
					</p>
					<div class="tif-acctdash__cards">
						<div class="tif-card" data-card="receivables_outstanding">
							<div class="tif-card__label">Receivables outstanding</div>
							<div class="tif-card__value">—</div>
							<div class="tif-card__hint">Sales Invoices (submitted)</div>
						</div>
						<div class="tif-card" data-card="receivables_overdue">
							<div class="tif-card__label">Receivables overdue</div>
							<div class="tif-card__value">—</div>
							<div class="tif-card__hint">Due date &lt; today</div>
						</div>
						<div class="tif-card" data-card="payables_outstanding">
							<div class="tif-card__label">Payables outstanding</div>
							<div class="tif-card__value">—</div>
							<div class="tif-card__hint">Purchase Invoices (submitted)</div>
						</div>
						<div class="tif-card" data-card="payables_overdue">
							<div class="tif-card__label">Payables overdue</div>
							<div class="tif-card__value">—</div>
							<div class="tif-card__hint">Due date &lt; today</div>
						</div>
						<div class="tif-card" data-card="income_ytd">
							<div class="tif-card__label">Income (YTD)</div>
							<div class="tif-card__value">—</div>
							<div class="tif-card__hint">GL Income</div>
						</div>
						<div class="tif-card" data-card="expense_ytd">
							<div class="tif-card__label">Expense (YTD)</div>
							<div class="tif-card__value">—</div>
							<div class="tif-card__hint">GL Expense</div>
						</div>
						<div class="tif-card" data-card="net_ytd">
							<div class="tif-card__label">Net (YTD)</div>
							<div class="tif-card__value">—</div>
							<div class="tif-card__hint">Income − Expense</div>
						</div>
					</div>

					<p class="text-muted" style="font-size:12px;margin:16px 0 8px">
						<strong>Operational</strong> — drafts that need action
					</p>
					<div class="tif-acctdash__cards">
						<div class="tif-card" data-card="draft_sales_invoices">
							<div class="tif-card__label">Draft Sales Invoices</div>
							<div class="tif-card__value">—</div>
							<div class="tif-card__hint">docstatus = 0</div>
						</div>
						<div class="tif-card" data-card="draft_purchase_invoices">
							<div class="tif-card__label">Draft Purchase Invoices</div>
							<div class="tif-card__value">—</div>
							<div class="tif-card__hint">docstatus = 0</div>
						</div>
						<div class="tif-card" data-card="draft_payment_entries">
							<div class="tif-card__label">Draft Payment Entries</div>
							<div class="tif-card__value">—</div>
							<div class="tif-card__hint">docstatus = 0</div>
						</div>
						<div class="tif-card" data-card="draft_journal_entries">
							<div class="tif-card__label">Draft Journal Entries</div>
							<div class="tif-card__value">—</div>
							<div class="tif-card__hint">docstatus = 0</div>
						</div>
					</div>

					<div class="tif-acctdash__grid">
						<div class="tif-panel tif-panel--span2">
							<div class="tif-panel__title">Monthly income vs expense</div>
							<div id="tif-acctdash-income-expense"></div>
						</div>
						<div class="tif-panel">
							<div class="tif-panel__title">Accounts team activity</div>
							<div class="tif-panel__hint">Created &amp; last touched in the selected date range</div>
							<div class="tif-table" data-table="accounts_user_activity"></div>
						</div>
						<div class="tif-panel">
							<div class="tif-panel__title">Bank / Cash balances</div>
							<div class="tif-panel__hint">Top accounts by absolute balance</div>
							<div class="tif-table" data-table="bank_cash_balances"></div>
						</div>
						<div class="tif-panel tif-panel--span2">
							<div class="tif-panel__title">Overdue receivables</div>
							<div class="tif-table" data-table="overdue_receivables"></div>
						</div>
						<div class="tif-panel tif-panel--span2">
							<div class="tif-panel__title">Overdue payables</div>
							<div class="tif-table" data-table="overdue_payables"></div>
						</div>
					</div>
				</div>
			`);
		}

		inject_styles() {
			if (document.getElementById("tif-acctdash-style")) return;
			const style = document.createElement("style");
			style.id = "tif-acctdash-style";
			style.textContent = `
				.tif-acctdash__cards{display:grid;grid-template-columns:repeat(4,minmax(0,1fr));gap:10px}
				@media (max-width: 1200px){.tif-acctdash__cards{grid-template-columns:repeat(2,minmax(0,1fr));}}
				.tif-card{border:1px solid var(--border-color,#e5e7eb);border-radius:10px;background:var(--card-bg,#fff);padding:12px}
				.tif-card__label{font-size:12px;color:var(--text-muted,#64748b);font-weight:600}
				.tif-card__value{font-size:20px;margin-top:6px;font-weight:700}
				.tif-card__hint{font-size:11px;color:var(--text-muted,#64748b);margin-top:4px}
				.tif-acctdash__grid{margin-top:14px;display:grid;grid-template-columns:repeat(2,minmax(0,1fr));gap:12px}
				@media (max-width: 1200px){.tif-acctdash__grid{grid-template-columns:1fr;}}
				.tif-panel{border:1px solid var(--border-color,#e5e7eb);border-radius:12px;background:var(--card-bg,#fff);padding:12px}
				.tif-panel--span2{grid-column:span 2}
				@media (max-width: 1200px){.tif-panel--span2{grid-column:auto}}
				.tif-panel__title{font-size:13px;font-weight:700;margin-bottom:8px}
				.tif-panel__hint{font-size:11px;color:var(--text-muted,#64748b);margin:-4px 0 10px}
				.tif-table table{width:100%;border-collapse:collapse}
				.tif-table th,.tif-table td{border-top:1px solid var(--border-color,#e5e7eb);padding:8px;font-size:12px}
				.tif-table th{color:var(--text-muted,#64748b);font-weight:700;text-align:left}
				.tif-table td.num{text-align:right;font-variant-numeric:tabular-nums}
				.tif-link{color:var(--text-color,#0f172a);text-decoration:underline;cursor:pointer}
			`;
			document.head.appendChild(style);
		}

		make_filters() {
			this.page.clear_actions && this.page.clear_actions();
			this.page.clear_custom_actions && this.page.clear_custom_actions();

			this.page.add_field({
				label: "Company",
				fieldtype: "Link",
				fieldname: "company",
				options: "Company",
				default: this.filters.company,
				change: () => {
					this.filters.company = this.page.fields_dict.company.get_value() || "";
					this.refresh();
				},
			});

			this.page.add_field({
				label: "From Date",
				fieldtype: "Date",
				fieldname: "from_date",
				default: this.filters.from_date,
				change: () => {
					this.filters.from_date = this.page.fields_dict.from_date.get_value() || null;
					this.refresh();
				},
			});

			this.page.add_field({
				label: "To Date",
				fieldtype: "Date",
				fieldname: "to_date",
				default: this.filters.to_date,
				change: () => {
					this.filters.to_date = this.page.fields_dict.to_date.get_value() || frappe.datetime.get_today();
					this.refresh();
				},
			});
		}

		load_apexcharts() {
			if (this.apex_loaded || window.ApexCharts) {
				this.apex_loaded = true;
				return Promise.resolve();
			}
			return new Promise((resolve, reject) => {
				const existing = document.querySelector('script[data-tif-apexcharts="1"]');
				if (existing) {
					existing.addEventListener("load", () => resolve());
					existing.addEventListener("error", () => reject());
					return;
				}
				const script = document.createElement("script");
				script.dataset.tifApexcharts = "1";
				script.src = "https://cdn.jsdelivr.net/npm/apexcharts@3.49.1";
				script.onload = () => {
					this.apex_loaded = true;
					resolve();
				};
				script.onerror = () => reject(new Error("Failed to load ApexCharts"));
				document.head.appendChild(script);
			});
		}

		set_card(key, value, fmt = false) {
			const el = this.page.main.find(`.tif-card[data-card="${key}"] .tif-card__value`)[0];
			if (!el) return;
			if (value === null || value === undefined || value === "") {
				el.textContent = "—";
				return;
			}
			let formatted;
			if (fmt === "currency") {
				formatted = frappe.format(value, { fieldtype: "Currency" });
			} else if (typeof value === "number") {
				formatted = frappe.format(value, { fieldtype: Number.isInteger(value) ? "Int" : "Currency" });
			} else {
				formatted = String(value);
			}
			const s = String(formatted ?? value);
			el.textContent = frappe.utils?.strip_html ? frappe.utils.strip_html(s) : s.replace(/<[^>]*>/g, "");
		}

		ensure_chart(key, selector, options) {
			if (!window.ApexCharts) return null;
			if (this.charts[key]) return this.charts[key];
			const container = this.page.main.find(selector)[0];
			if (!container) return null;
			const chart = new ApexCharts(container, options);
			chart.render();
			this.charts[key] = chart;
			return chart;
		}

		render_table(key, rows, columns) {
			const container = this.page.main.find(`.tif-table[data-table="${key}"]`)[0];
			if (!container) return;
			const safeRows = rows || [];
			if (!safeRows.length) {
				container.innerHTML = `<div class="text-muted" style="font-size:12px">No data</div>`;
				return;
			}
			const thead = `<thead><tr>${columns.map((c) => `<th>${frappe.utils.escape_html(c.label)}</th>`).join("")}</tr></thead>`;
			const tbody = safeRows
				.map((r) => {
					return `<tr>${columns
						.map((c) => {
							const v = r?.[c.key];
							if (c.type === "currency") {
								const formatted = frappe.format(v || 0, { fieldtype: "Currency" });
								const stripped = frappe.utils?.strip_html
									? frappe.utils.strip_html(String(formatted))
									: String(formatted).replace(/<[^>]*>/g, "");
								return `<td class="num">${frappe.utils.escape_html(stripped)}</td>`;
							}
							if (c.type === "link") {
								const doctype = c.doctype || (c.doctype_key ? r?.[c.doctype_key] : null);
								const label = v || r?.[c.label_key] || "";
								if (!doctype || !v) {
									return `<td>${frappe.utils.escape_html(String(label || ""))}</td>`;
								}
								return `<td><span class="tif-link" data-doctype="${frappe.utils.escape_html(
									String(doctype)
								)}" data-name="${frappe.utils.escape_html(String(v))}">${frappe.utils.escape_html(
									String(label)
								)}</span></td>`;
							}
							return `<td>${frappe.utils.escape_html(String(v ?? ""))}</td>`;
						})
						.join("")}</tr>`;
				})
				.join("");
			container.innerHTML = `<table>${thead}<tbody>${tbody}</tbody></table>`;
			container.querySelectorAll(".tif-link").forEach((el) => {
				el.addEventListener("click", () => {
					const doctype = el.getAttribute("data-doctype");
					const name = el.getAttribute("data-name");
					if (doctype && name) frappe.set_route("Form", doctype, name);
				});
			});
		}

		async refresh() {
			if (!window.ApexCharts) {
				try {
					await this.load_apexcharts();
				} catch (e) {
					frappe.msgprint("ApexCharts could not be loaded.");
					return;
				}
			}
			const filters = { ...this.filters };
			this.page.set_indicator("Loading…", "blue");
			const r = await frappe.call({
				method: "tif_customization.tif_customization.page.accounts_status.accounts_status.get_dashboard_data",
				args: { filters },
			});
			if (this.page.clear_indicator) this.page.clear_indicator();
			const data = r?.message || {};
			this.render_kpis(data);
			this.render_charts(data);
			this.render_tables(data);
		}

		render_kpis(data) {
			this.set_card("receivables_outstanding", data.receivables_outstanding, "currency");
			this.set_card("receivables_overdue", data.receivables_overdue, "currency");
			this.set_card("payables_outstanding", data.payables_outstanding, "currency");
			this.set_card("payables_overdue", data.payables_overdue, "currency");
			this.set_card("income_ytd", data.income_ytd, "currency");
			this.set_card("expense_ytd", data.expense_ytd, "currency");
			this.set_card("net_ytd", data.net_ytd, "currency");
			this.set_card("draft_sales_invoices", data.draft_sales_invoices);
			this.set_card("draft_purchase_invoices", data.draft_purchase_invoices);
			this.set_card("draft_payment_entries", data.draft_payment_entries);
			this.set_card("draft_journal_entries", data.draft_journal_entries);
		}

		render_charts(data) {
			const mode = frappe.get_cookie("theme") === "dark" ? "dark" : "light";
			const d = data.monthly_income_expense || {};
			const chart = this.ensure_chart("income_expense", "#tif-acctdash-income-expense", {
				chart: { type: "line", height: 320, toolbar: { show: false }, zoom: { enabled: false } },
				theme: { mode },
				colors: TIF_ACCT_PALETTE.lines2,
				stroke: { width: 3, curve: "smooth" },
				markers: { size: 3, strokeWidth: 2 },
				grid: { borderColor: "#e2e8f0", strokeDashArray: 4 },
				tooltip: { theme: mode, shared: true, intersect: false },
				noData: { text: "No data" },
				xaxis: { categories: d.labels || [] },
				series: d.series || [],
				yaxis: { labels: { formatter: (v) => frappe.format(v, { fieldtype: "Currency" }) } },
			});
			if (chart) {
				chart.updateOptions({ xaxis: { categories: d.labels || [] }, series: d.series || [] }, false, true);
			}
		}

		render_tables(data) {
			this.render_table("accounts_user_activity", data.accounts_user_activity, [
				{ key: "user", label: "User" },
				{ key: "created_count", label: "Created" },
				{ key: "touched_count", label: "Touched" },
				{ key: "last_name", label: "Last entry", type: "link", doctype_key: "last_doctype" },
				{ key: "last_action_at", label: "Last at" },
			]);
			this.render_table("bank_cash_balances", data.bank_cash_balances, [
				{ key: "account", label: "Account" },
				{ key: "balance", label: "Balance", type: "currency" },
			]);
			this.render_table("overdue_receivables", data.overdue_receivables, [
				{ key: "name", label: "Invoice", type: "link", doctype: "Sales Invoice" },
				{ key: "customer", label: "Customer" },
				{ key: "due_date", label: "Due" },
				{ key: "outstanding_amount", label: "Outstanding", type: "currency" },
			]);
			this.render_table("overdue_payables", data.overdue_payables, [
				{ key: "name", label: "Bill", type: "link", doctype: "Purchase Invoice" },
				{ key: "supplier", label: "Supplier" },
				{ key: "due_date", label: "Due" },
				{ key: "outstanding_amount", label: "Outstanding", type: "currency" },
			]);
		}
	}

	wrapper.tifAccountsStatusDashboard = new TIFAccountsStatusDashboard(page);
	wrapper.tifAccountsStatusDashboard.make();
};

File: tif_customization/tif_customization/page/accounts_status/accounts_status.py
import calendar
import json

import frappe
from frappe.utils import add_months, cint, flt, get_first_day, get_last_day, getdate, nowdate


@frappe.whitelist()
def get_dashboard_data(filters=None):
	"""Trustee-friendly finance monitoring KPIs for Accounts Status page."""
	if isinstance(filters, str):
		filters = json.loads(filters or "{}")
	filters = filters or {}

	_allowed_roles = {"System Manager", "Accounts Manager", "Accounts User", "Trustee"}
	if not (set(frappe.get_roles()) & _allowed_roles):
		frappe.throw("Not permitted", frappe.PermissionError)

	company = (filters.get("company") or frappe.defaults.get_user_default("Company") or "").strip()
	from_date, to_date = _date_range(filters, company)

	out = _empty_payload(company, from_date, to_date)

	# Core KPIs (respect DocType permissions by falling back safely on error)
	out.update(_ar_kpis(company, to_date))
	out.update(_ap_kpis(company, to_date))
	out.update(_draft_voucher_counts(company))
	out.update(_ytd_income_expense(company, from_date, to_date))

	out["monthly_income_expense"] = _monthly_income_expense(company, from_date, to_date)
	out["bank_cash_balances"] = _bank_cash_balances(company, to_date, limit=25)
	out["overdue_receivables"] = _overdue_receivables_rows(company, to_date, limit=20)
	out["overdue_payables"] = _overdue_payables_rows(company, to_date, limit=20)
	out["accounts_user_activity"] = _accounts_user_activity(company, from_date, to_date)

	return out


def _empty_payload(company, from_date, to_date):
	return {
		"filters": {"company": company, "from_date": str(from_date), "to_date": str(to_date)},
		"receivables_outstanding": 0.0,
		"receivables_overdue": 0.0,
		"receivables_overdue_count": 0,
		"payables_outstanding": 0.0,
		"payables_overdue": 0.0,
		"payables_overdue_count": 0,
		"draft_sales_invoices": 0,
		"draft_purchase_invoices": 0,
		"draft_payment_entries": 0,
		"draft_journal_entries": 0,
		"income_ytd": 0.0,
		"expense_ytd": 0.0,
		"net_ytd": 0.0,
		"monthly_income_expense": {"labels": [], "series": []},
		"bank_cash_balances": [],
		"overdue_receivables": [],
		"overdue_payables": [],
		"accounts_user_activity": [],
	}


def _has_field(doctype, fieldname):
	try:
		return bool(frappe.get_meta(doctype).get_field(fieldname))
	except Exception:
		return False


def _date_range(filters, company):
	today = getdate(nowdate())
	from_date = filters.get("from_date")
	to_date = filters.get("to_date")
	if from_date:
		from_date = getdate(from_date)
	if to_date:
		to_date = getdate(to_date)
	if not to_date:
		to_date = today
	if not from_date:
		from_date = _fiscal_year_start(to_date, company)
	if from_date > to_date:
		from_date, to_date = to_date, from_date
	return from_date, to_date


def _fiscal_year_start(dt, company):
	try:
		from erpnext.accounts.utils import get_fiscal_year

		_, fy_start, _ = get_fiscal_year(dt, company=company)
		return getdate(fy_start)
	except Exception:
		# Safe fallback: calendar year
		return getdate(f"{dt.year}-01-01")


def _ar_kpis(company, to_date):
	out = {"receivables_outstanding": 0.0, "receivables_overdue": 0.0, "receivables_overdue_count": 0}
	if not frappe.db.table_exists("Sales Invoice"):
		return out
	if not company or not _has_field("Sales Invoice", "company"):
		return out
	try:
		row = frappe.db.sql(
			"""
			SELECT
				COALESCE(SUM(si.outstanding_amount), 0) AS outstanding,
				COALESCE(SUM(CASE WHEN si.due_date IS NOT NULL AND si.due_date < %(to_date)s THEN si.outstanding_amount ELSE 0 END), 0) AS overdue,
				COALESCE(SUM(CASE WHEN si.due_date IS NOT NULL AND si.due_date < %(to_date)s THEN 1 ELSE 0 END), 0) AS overdue_count
			FROM `tabSales Invoice` si
			WHERE si.docstatus = 1
			  AND si.company = %(company)s
			  AND si.outstanding_amount > 0
			""",
			{"company": company, "to_date": str(to_date)},
			as_dict=True,
		)
		r = (row or [{}])[0]
		out["receivables_outstanding"] = flt(r.get("outstanding"))
		out["receivables_overdue"] = flt(r.get("overdue"))
		out["receivables_overdue_count"] = cint(r.get("overdue_count"))
	except Exception:
		pass
	return out


def _ap_kpis(company, to_date):
	out = {"payables_outstanding": 0.0, "payables_overdue": 0.0, "payables_overdue_count": 0}
	if not frappe.db.table_exists("Purchase Invoice"):
		return out
	if not company or not _has_field("Purchase Invoice", "company"):
		return out
	try:
		row = frappe.db.sql(
			"""
			SELECT
				COALESCE(SUM(pi.outstanding_amount), 0) AS outstanding,
				COALESCE(SUM(CASE WHEN pi.due_date IS NOT NULL AND pi.due_date < %(to_date)s THEN pi.outstanding_amount ELSE 0 END), 0) AS overdue,
				COALESCE(SUM(CASE WHEN pi.due_date IS NOT NULL AND pi.due_date < %(to_date)s THEN 1 ELSE 0 END), 0) AS overdue_count
			FROM `tabPurchase Invoice` pi
			WHERE pi.docstatus = 1
			  AND pi.company = %(company)s
			  AND pi.outstanding_amount > 0
			""",
			{"company": company, "to_date": str(to_date)},
			as_dict=True,
		)
		r = (row or [{}])[0]
		out["payables_outstanding"] = flt(r.get("outstanding"))
		out["payables_overdue"] = flt(r.get("overdue"))
		out["payables_overdue_count"] = cint(r.get("overdue_count"))
	except Exception:
		pass
	return out


def _draft_voucher_counts(company):
	out = {
		"draft_sales_invoices": 0,
		"draft_purchase_invoices": 0,
		"draft_payment_entries": 0,
		"draft_journal_entries": 0,
	}
	params = {"company": company}
	if company:
		company_cond = " AND company = %(company)s "
	else:
		company_cond = ""

	def _count(doctype, extra_where=""):
		if not frappe.db.table_exists(doctype):
			return 0
		if company and not _has_field(doctype, "company"):
			return 0
		try:
			row = frappe.db.sql(
				f"SELECT COUNT(name) AS c FROM `tab{doctype}` WHERE docstatus = 0 {company_cond} {extra_where}",
				params if company else {},
				as_dict=True,
			)
			return cint((row or [{}])[0].get("c"))
		except Exception:
			return 0

	out["draft_sales_invoices"] = _count("Sales Invoice")
	out["draft_purchase_invoices"] = _count("Purchase Invoice")
	out["draft_payment_entries"] = _count("Payment Entry")
	out["draft_journal_entries"] = _count("Journal Entry")
	return out


def _ytd_income_expense(company, from_date, to_date):
	out = {"income_ytd": 0.0, "expense_ytd": 0.0, "net_ytd": 0.0}
	if not frappe.db.table_exists("GL Entry") or not frappe.db.table_exists("Account"):
		return out
	if not company:
		return out
	try:
		row = frappe.db.sql(
			"""
			SELECT
				COALESCE(SUM(CASE WHEN acc.root_type = 'Income' THEN (gle.credit - gle.debit) ELSE 0 END), 0) AS income,
				COALESCE(SUM(CASE WHEN acc.root_type = 'Expense' THEN (gle.debit - gle.credit) ELSE 0 END), 0) AS expense
			FROM `tabGL Entry` gle
			JOIN `tabAccount` acc ON acc.name = gle.account
			WHERE gle.is_cancelled = 0
			  AND gle.company = %(company)s
			  AND gle.posting_date BETWEEN %(from_date)s AND %(to_date)s
			  AND acc.root_type IN ('Income', 'Expense')
			""",
			{"company": company, "from_date": str(from_date), "to_date": str(to_date)},
			as_dict=True,
		)
		r = (row or [{}])[0]
		out["income_ytd"] = flt(r.get("income"))
		out["expense_ytd"] = flt(r.get("expense"))
		out["net_ytd"] = flt(out["income_ytd"]) - flt(out["expense_ytd"])
	except Exception:
		pass
	return out


def _monthly_income_expense(company, from_date, to_date):
	labels = []
	income = []
	expense = []
	if not company or not frappe.db.table_exists("GL Entry") or not frappe.db.table_exists("Account"):
		return {"labels": labels, "series": []}

	cur = get_first_day(from_date)
	while cur <= to_date:
		m_from = get_first_day(cur)
		m_to = min(get_last_day(cur), to_date)
		labels.append(cur.strftime("%b %Y"))
		res = _ytd_income_expense(company, m_from, m_to)
		income.append(flt(res.get("income_ytd")))
		expense.append(flt(res.get("expense_ytd")))
		cur = add_months(cur, 1)

	return {"labels": labels, "series": [{"name": "Income", "data": income}, {"name": "Expense", "data": expense}]}


def _bank_cash_balances(company, to_date, limit=25):
	if not company or not frappe.db.table_exists("GL Entry") or not frappe.db.table_exists("Account"):
		return []
	if not _has_field("Account", "account_type"):
		return []
	try:
		rows = frappe.db.sql(
			f"""
			SELECT
				acc.name AS account,
				COALESCE(NULLIF(TRIM(acc.account_name), ''), acc.name) AS account_name,
				COALESCE(SUM(gle.debit - gle.credit), 0) AS balance
			FROM `tabAccount` acc
			LEFT JOIN `tabGL Entry` gle
				ON gle.account = acc.name
				AND gle.is_cancelled = 0
				AND gle.company = %(company)s
				AND gle.posting_date <= %(to_date)s
			WHERE acc.company = %(company)s
			  AND acc.account_type IN ('Bank', 'Cash')
			GROUP BY acc.name, acc.account_name
			ORDER BY ABS(balance) DESC
			LIMIT {cint(limit)}
			""",
			{"company": company, "to_date": str(to_date)},
			as_dict=True,
		)
		return [
			{"account": r.get("account"), "account_name": r.get("account_name"), "balance": flt(r.get("balance"))}
			for r in (rows or [])
		]
	except Exception:
		return []


def _overdue_receivables_rows(company, to_date, limit=20):
	if not company or not frappe.db.table_exists("Sales Invoice") or not _has_field("Sales Invoice", "company"):
		return []
	fields = ["name", "customer", "posting_date", "due_date", "outstanding_amount", "grand_total"]
	available = [f for f in fields if _has_field("Sales Invoice", f)]
	if not available:
		return []
	try:
		rows = frappe.db.sql(
			f"""
			SELECT {", ".join([f"si.`{f}`" for f in available])}
			FROM `tabSales Invoice` si
			WHERE si.docstatus = 1
			  AND si.company = %(company)s
			  AND si.outstanding_amount > 0
			  AND si.due_date IS NOT NULL
			  AND si.due_date < %(to_date)s
			ORDER BY si.due_date ASC, si.outstanding_amount DESC
			LIMIT {cint(limit)}
			""",
			{"company": company, "to_date": str(to_date)},
			as_dict=True,
		)
		return rows or []
	except Exception:
		return []


def _overdue_payables_rows(company, to_date, limit=20):
	if not company or not frappe.db.table_exists("Purchase Invoice") or not _has_field("Purchase Invoice", "company"):
		return []
	fields = ["name", "supplier", "posting_date", "due_date", "outstanding_amount", "grand_total"]
	available = [f for f in fields if _has_field("Purchase Invoice", f)]
	if not available:
		return []
	try:
		rows = frappe.db.sql(
			f"""
			SELECT {", ".join([f"pi.`{f}`" for f in available])}
			FROM `tabPurchase Invoice` pi
			WHERE pi.docstatus = 1
			  AND pi.company = %(company)s
			  AND pi.outstanding_amount > 0
			  AND pi.due_date IS NOT NULL
			  AND pi.due_date < %(to_date)s
			ORDER BY pi.due_date ASC, pi.outstanding_amount DESC
			LIMIT {cint(limit)}
			""",
			{"company": company, "to_date": str(to_date)},
			as_dict=True,
		)
		return rows or []
	except Exception:
		return []


def _accounts_user_activity(company, from_date, to_date):
	"""Activity summary for accounts users (counts + last touch) within selected range."""
	users = [
		"muhammad.yasir@tif.edu.pk",
		"muhammad.raza@tif.edu.pk",
	]

	doctypes = [
		"Sales Invoice",
		"Purchase Invoice",
		"Payment Entry",
		"Journal Entry",
	]

	out = []
	for user in users:
		row = {
			"user": user,
			"created_count": 0,
			"touched_count": 0,
			"last_doctype": "",
			"last_name": "",
			"last_action_at": "",
			"last_docstatus": None,
		}
		last = None
		for dt in doctypes:
			if not frappe.db.table_exists(dt):
				continue
			company_cond = ""
			params = {"u": user, "from_date": str(from_date), "to_date": str(to_date)}
			if company and _has_field(dt, "company"):
				company_cond = " AND t.company = %(company)s "
				params["company"] = company

			try:
				created = frappe.db.sql(
					f"""
					SELECT COUNT(*) AS c
					FROM `tab{dt}` t
					WHERE t.owner = %(u)s
					  AND t.creation BETWEEN %(from_date)s AND %(to_date)s
					  {company_cond}
					""",
					params,
					as_dict=True,
				)
				row["created_count"] += cint((created or [{}])[0].get("c"))
			except Exception:
				pass

			try:
				touched = frappe.db.sql(
					f"""
					SELECT COUNT(*) AS c
					FROM `tab{dt}` t
					WHERE t.modified_by = %(u)s
					  AND t.modified BETWEEN %(from_date)s AND %(to_date)s
					  {company_cond}
					""",
					params,
					as_dict=True,
				)
				row["touched_count"] += cint((touched or [{}])[0].get("c"))
			except Exception:
				pass

			try:
				last_rows = frappe.db.sql(
					f"""
					SELECT t.name AS name, t.modified AS modified, t.docstatus AS docstatus
					FROM `tab{dt}` t
					WHERE t.modified_by = %(u)s
					  {company_cond}
					ORDER BY t.modified DESC
					LIMIT 1
					""",
					params if company else {"u": user},
					as_dict=True,
				)
				cand = (last_rows or [None])[0]
				if cand and (not last or (cand.get("modified") and cand.get("modified") > last.get("modified"))):
					last = {"doctype": dt, **cand}
			except Exception:
				pass

		if last:
			row["last_doctype"] = last.get("doctype") or ""
			row["last_name"] = last.get("name") or ""
			row["last_action_at"] = str(last.get("modified") or "")
			row["last_docstatus"] = last.get("docstatus")

		out.append(row)

	return out
